File: examples/wolfspeed/TransientThermal.cpp
#include <boost/stacktrace.hpp>
#include <string_view>
#include <filesystem>
#include <cassert>
#include <csignal>

#include "Common.hpp"
#include "../test/TestData.hpp"
#include "generic/tools/StringHelper.hpp"
#include "utility/ELayoutRetriever.h"
#include "EDataMgr.h"

void SignalHandler(int signum)
{
    ::signal(signum, SIG_DFL);
    std::cout << boost::stacktrace::stacktrace();
    ::raise(SIGABRT);
}

using namespace ecad;
using namespace generic;
auto & eDataMgr = EDataMgr::Instance();

std::vector<FPoint3D> GetDieMonitors(CPtr<ILayoutView> layout)
{
    std::vector<FPoint3D> monitors;
    utils::ELayoutRetriever retriever(layout);
    std::vector<std::string> cellInsts{"TopBridge1", "TopBridge2", "BotBridge1", "BotBridge2"};
    std::vector<std::string> components{"Die1", "Die2", "Die3"};
    for(const auto & cellInst : cellInsts) {
        for (const auto & component : components) {
            std::string name = cellInst + eDataMgr.HierSep() + component;
            auto comp = layout->FindComponentByName(name);
            ECAD_ASSERT(comp != nullptr);
            EFloat elevation, thickness;
            retriever.GetComponentHeightThickness(comp, elevation, thickness);
            auto bonds = comp->GetBoundary(); { ECAD_ASSERT(bonds); }
            auto loc = generic::geometry::Extent(bonds->GetContour()).Center();
            monitors.emplace_back(loc[0], loc[1], elevation - 0.1 * thickness);
        }
    }
    return monitors;
}

void TransientThermalFlow(Ptr<ILayoutView> layout, const std::string & workDir, EFloat period, EFloat duty)
{
    auto extractionSettings = ExtractionSettings(workDir);

    EThermalTransientSimulationSetup setup(workDir, eDataMgr.Threads(), {});
    setup.monitors = GetDieMonitors(layout);
    setup.settings.envTemperature = {25, ETemperatureUnit::Celsius};
    setup.settings.mor.order = 0;
    setup.settings.mor.romLoadFile = workDir + ECAD_SEPS + "rom_source.db";
    setup.settings.mor.romSaveFile = setup.settings.mor.romLoadFile;
    setup.settings.verbose = true;
    setup.settings.dumpResults = true;
    setup.settings.duration = 10;
    setup.settings.step = period * duty / 10;
    setup.settings.temperatureDepend = false;
    setup.settings.samplingWindow = setup.settings.duration;
    setup.settings.minSamplingInterval = period * duty / 10;
    setup.settings.absoluteError = 1e-5;
    setup.settings.relativeError = 1e-5;
    setup.extractionSettings = std::move(extractionSettings);
    EThermalTransientExcitation excitation = [period, duty](EFloat t, size_t scen) -> EFloat { 
        EFloat tm = std::fmod(t, period); 
        switch (scen) {
            case 0 : return 1;
            case 1 : return tm < duty * period ? 1 : 0;
            case 2 : return 0.5 * period < tm && tm < (duty + 0.5) * period ? 1 : 0; 
            default : { ECAD_ASSERT(false) return 0; }
        }
    };
    auto [minT, maxT] = layout->RunThermalSimulation(setup, excitation);    
    ECAD_TRACE("minT: %1%, maxT: %2%", minT, maxT);
}

int main(int argc, char * argv[])
{
    ::signal(SIGSEGV, &SignalHandler);
    ::signal(SIGABRT, &SignalHandler);

    eDataMgr.Init(ecad::ELogLevel::Trace);

    std::string filename = generic::fs::DirName(__FILE__).string() + "/data/design/CAS300M12BM2.ecad";
    auto database = eDataMgr.LoadDatabase(filename);
    if (nullptr == database) return EXIT_FAILURE;
    auto cell = database->FindCellByName("Base");
    std::string workDir = generic::fs::DirName(__FILE__).string() + "data/simulation/transient";
    auto layout = cell->GetFlattenedLayoutView();
    TransientThermalFlow(layout, workDir, 0.02, 0.5);
    ecad::EDataMgr::Instance().ShutDown();
    return EXIT_SUCCESS;
}

File: examples/python/wolfspeed/TransientThermal.py
import os
import sys
import SetupDesign
import StaticThermal
from SetupDesign import ecad

mgr = ecad.DataMgr

def get_simulation_setup(layout, work_dir, cell_insts, components, force_rebuild = False) :

    extraction_setting = StaticThermal.get_extraction_setting(work_dir, force_rebuild)
    extraction_setting.layer_cut_settings.layer_transition_ratio = 5
    simulation_setup = ecad.ThermalTransientSimulationSetup(work_dir, mgr.threads(), set())
    simulation_setup.set_extraction_settings(extraction_setting)
    simulation_setup.settings.env_temperature = ecad.Temperature(25, ecad.TemperatureUnit.CELSIUS)
    simulation_setup.settings.mor.order = 0
    simulation_setup.settings.verbose = True
    simulation_setup.settings.dump_results = True
    simulation_setup.settings.duration = 100
    simulation_setup.settings.step = 10
    simulation_setup.settings.temperature_depend = False
    simulation_setup.settings.sampling_window = 100
    simulation_setup.settings.min_sampling_interval = 0.01
    simulation_setup.settings.absolute_error = 1e-2
    simulation_setup.settings.relative_error = 1e-3
    simulation_setup.settings.adaptive = True
    simulation_setup.monitors = StaticThermal.get_die_monitors(layout, cell_insts, components)
    return simulation_setup

def transient_thermal_flow(chip_locations, include_joule_heat = True) :
    
    work_dir = os.path.dirname(__file__) + '/data/simulation/transient'
    layout = SetupDesign.setup_design("CAS300M12BM2", chip_locations, work_dir)
    if not layout :
        return 0, 0, []

    StaticThermal.setup_power(layout, include_joule_heat)
    setup = get_simulation_setup(layout, work_dir, ["TopBridge1", "TopBridge2", "BotBridge1", "BotBridge2"], ["Die1", "Die2", "Die3"])
    
    def excitation(t, scen) :
        return 1 #const power
    results = layout.run_thermal_simulation(setup, excitation)
    return results
    
def main() :

    mgr.init(ecad.LogLevel.TRACE)
    transient_thermal_flow(SetupDesign.chip_locations, include_joule_heat=True)
    mgr.shut_down()

if __name__ == '__main__' :
    main()